substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -1980,6 +1983,115 @@ pub mod pallet {
 
 			Ok(Some(T::WeightInfo::deprecate_controller_batch(controllers.len() as u32)).into())
 		}
+
+		/// Restores the state of a ledger which is in an inconsistent state.
+		///
+		/// The requirements to restore a ledger are the following:
+		/// * The stash is bonded;
+		/// * If the stash has an associated ledger, its state must be inconsistent;
+		/// * If the ledger is not corrupted *but* its staking lock is out of sync.
+		///
+		/// The `maybe_*` input parameters will overwrite the corresponding data and metadata of the
+		/// ledger associated with the stash. If the input parameters are not set, the ledger will
+		/// be reset with the current values.
+		///
+		/// Upon successful execution, this extrinsic will *recreate* the ledger based on its past
+		/// state and/or the `maybe_controller` and `maybe_total` input parameters.
+		#[pallet::call_index(29)]
+		#[pallet::weight(T::WeightInfo::restore_ledger())]
+		pub fn restore_ledger(
+			origin: OriginFor<T>,
+			stash: T::AccountId,
+			maybe_controller: Option<T::AccountId>,
+			maybe_total: Option<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			// closure to call before return from extrinsic to ensure the restore was successful.
+			let integrity_checks = || {
+				let lock = T::Currency::balance_locked(crate::STAKING_ID, &stash);
+
+				if let Some(ledger) = Bonded::<T>::get(&stash).and_then(Ledger::<T>::get) {
+					ensure!(lock == ledger.total, Error::<T>::BadState);
+					ensure!(Payee::<T>::get(&stash).is_some(), Error::<T>::BadState);
+					ensure!(ledger.stash == stash, Error::<T>::BadState);
+				} else {
+					ensure!(lock == Zero::zero(), Error::<T>::BadState);
+				}
+
+				Ok::<_, Error<T>>(())
+			};
+
+			let current_lock = T::Currency::balance_locked(crate::STAKING_ID, &stash);
+
+			let (new_controller, new_total) = match Self::inspect_bond_state(&stash) {
+				Ok(LedgerIntegrityState::Corrupted) => {
+					let new_controller = maybe_controller.unwrap_or(stash.clone());
```

Comment:
This controller should not be same as the last controller right? Practically we cannot just insert a new controller here. I believe the only option here is to make controller same as stash?

Reply:
This is true, and in the [recovery cases I tested](https://hackmd.io/v-XNrEoGRpe7APR-EZGhOA?view), we don't need to set the controller for any of the ledgers. I kept the convention that we may be able to explicitly set all the items of the ledger through this extrinsic (and lock). This argument is not the strongest and I'm happy to reconsider it and remove the `maybe_controller`.